Make more tests robust & independent

Unfortunately, even though #784 improved the tests a lot, we still seem to see flaky behaviour. So let's continue the same kind of work from #784 on these tests:

- [ ]
  ```python
  FAILED message_ix/tests/test_feature_bound_activity_shares.py::test_add_share_mode_lo - assert False
   +  where False = <function isclose at 0x111d84700>(0.5416666666666666, 0.525)
   +    where <function isclose at 0x111d84700> = np.isclose
  ```

Another one bites the dust:

- [ ] 
  ```python
  FAILED message_ix/tests/test_cli.py::test_copy_model - AssertionError: assert 1 == 0
   +  where 1 = <Result FileNotFoundError(2, 'No such file or directory')>.exit_code
  On: macos-latest-py3.9
  ```

Another flaky test appeared today: 
- [ ]
  ```python
   FAILED message_ix/tests/test_util.py::test_testing_make_scenario - RuntimeError: unhandled Java exception: The index set 'emission' does not have an element 'CO2'!
  On: macos-13-py3.10
  ```

There goes the next one:

- [ ]
  ```python
  FAILED message_ix/tests/test_feature_bound_activity_shares.py::test_add_share_mode_lo - assert False
   +  where False = <function isclose at 0x000001BF5AF800F0>(0.5416666666666666, 0.525)
   +    where <function isclose at 0x000001BF5AF800F0> = np.isclose
  On: windows-latest-py3.11
  ```

- [ ]
  ```python
   FAILED message_ix/tests/test_feature_bound_activity_shares.py::test_commodity_share_lo - assert False
   +  where False = <function isclose at 0x10fe1d8b0>(0.0, 0.15384615384615385)
   +    where <function isclose at 0x10fe1d8b0> = np.isclose
   +    and   0.15384615384615385 = <function test_commodity_share_lo.<locals>.calc_share at 0x161a71ca0>(<message_ix.core.Scenario object at 0x158b22370>)
  On: macos-latest-py3.8
  ```

--- a/RELEASE_NOTES.rst
+++ b/RELEASE_NOTES.rst
@@ -7,8 +7,10 @@ GitHub community guidelines
 Add community guidelines for interaction on GitHub (:pull:`871`).
 Please familiarize yourself with these to foster an open and welcoming community!
 
-.. All changes
-.. -----------
+All changes
+-----------
+
+- Add option to :func:`.util.copy_model` from a non-default location of model files (:pull:`877`).
 
 .. _v3.9.0:
 

--- a/doc/install-adv.rst
+++ b/doc/install-adv.rst
@@ -46,7 +46,7 @@ Common ways to install Java include:
 - Use a version of Java bundled or packaged for your operating system (for example, Ubuntu and other Linux distributions).
 
 If using Anaconda or Miniconda, installing Java manually is *not required*.
-Tis is because the ``message-ix`` conda-forge package depends on the `openjdk <https://anaconda.org/conda-forge/openjdk>`_ package, so the latter is automatically installed with the former.
+This is because the ``message-ix`` conda-forge package depends on the `openjdk <https://anaconda.org/conda-forge/openjdk>`_ package, so the latter is automatically installed with the former.
 
 
 .. _install-gams:

--- a/message_ix/cli.py
+++ b/message_ix/cli.py
@@ -34,10 +34,10 @@ ixmp.cli.ScenarioClass = message_ix.Scenario
 @click.option(
     "--source-dir",
     type=click.Path(file_okay=False),
-    help="Optional directory to copy file from.",
+    help="Optional directory to copy files from.",
 )
 @click.argument("path", type=click.Path(file_okay=False))
-def copy_model_cmd(path, source_dir, overwrite, set_default, quiet):
+def copy_model_cmd(path, overwrite, set_default, quiet, source_dir):
     """Copy the MESSAGE GAMS files to a new PATH.
 
     To use an existing set of GAMS files, you can also call:
@@ -46,7 +46,7 @@ def copy_model_cmd(path, source_dir, overwrite, set_default, quiet):
     """
     from message_ix.util import copy_model
 
-    copy_model(path, source_dir, overwrite, set_default, quiet)
+    copy_model(path, overwrite, set_default, quiet, source_dir=source_dir)
 
 
 @main.command()

--- a/message_ix/cli.py
+++ b/message_ix/cli.py
@@ -65,7 +65,7 @@ def dl(branch, tag, path):
     from urllib.request import Request, urlopen
 
     if tag and branch:
-        raise click.BadOptionUsage("Can only provide one of `tag` or `branch`")
+        raise click.UsageError(message="Can only provide one of `tag` or `branch`")
     elif branch:
         # Construct URL and filename from branch
         zipname = f"{branch}.zip"

--- a/message_ix/cli.py
+++ b/message_ix/cli.py
@@ -31,8 +31,13 @@ ixmp.cli.ScenarioClass = message_ix.Scenario
 )
 @click.option("--overwrite", is_flag=True, help="Overwrite existing files.")
 @click.option("--quiet", is_flag=True, help="Don't print paths of copied files.")
+@click.option(
+    "--source-dir",
+    type=click.Path(file_okay=False),
+    help="Optional directory to copy file from.",
+)
 @click.argument("path", type=click.Path(file_okay=False))
-def copy_model_cmd(path, overwrite, set_default, quiet):
+def copy_model_cmd(path, source_dir, overwrite, set_default, quiet):
     """Copy the MESSAGE GAMS files to a new PATH.
 
     To use an existing set of GAMS files, you can also call:
@@ -41,7 +46,7 @@ def copy_model_cmd(path, overwrite, set_default, quiet):
     """
     from message_ix.util import copy_model
 
-    copy_model(path, overwrite, set_default, quiet)
+    copy_model(path, source_dir, overwrite, set_default, quiet)
 
 
 @main.command()

--- a/message_ix/tests/conftest.py
+++ b/message_ix/tests/conftest.py
@@ -33,7 +33,12 @@ def message_ix_cli(tmp_env):
     class Runner(CliRunner):
         def invoke(self, *args, **kwargs):
             return super().invoke(
-                cli.main, list(filter(None, args)), env=tmp_env, **kwargs
+                cli.main,
+                list(filter(None, args)),
+                env=tmp_env,
+                catch_exceptions=False,
+                color=True,
+                **kwargs,
             )
 
     yield Runner().invoke

--- a/message_ix/tests/conftest.py
+++ b/message_ix/tests/conftest.py
@@ -40,9 +40,9 @@ def message_ix_cli(tmp_env):
 
 
 @pytest.fixture(scope="class")
-def message_test_mp(test_mp, request):
-    make_dantzig(test_mp, request=request)
-    make_dantzig(test_mp, multi_year=True, request=request)
+def message_test_mp(test_mp):
+    make_dantzig(test_mp)
+    make_dantzig(test_mp, multi_year=True)
     yield test_mp
 
 

--- a/message_ix/tests/conftest.py
+++ b/message_ix/tests/conftest.py
@@ -40,9 +40,9 @@ def message_ix_cli(tmp_env):
 
 
 @pytest.fixture(scope="class")
-def message_test_mp(test_mp):
-    make_dantzig(test_mp)
-    make_dantzig(test_mp, multi_year=True)
+def message_test_mp(test_mp, request):
+    make_dantzig(test_mp, request=request)
+    make_dantzig(test_mp, multi_year=True, request=request)
     yield test_mp
 
 

--- a/message_ix/tests/test_cli.py
+++ b/message_ix/tests/test_cli.py
@@ -49,36 +49,32 @@ def test_copy_model(
 
 
 @pytest.mark.parametrize(
-    "opts",
+    "opts, exit_code",
     [
-        "",
-        "--branch=main",
-        "--tag=v1.2.0",
+        ([""], 0),
+        (["--branch=main"], 0),
+        (["--tag=v1.2.0"], 0),
         # Nonexistent tag
-        pytest.param("--tag=v999", marks=pytest.mark.xfail(raises=ValueError)),
+        pytest.param(["--tag=v999"], 0, marks=pytest.mark.xfail(raises=ValueError)),
         # Can't use both --tag and --branch
+        # TODO Why is this not actually raising a click.UsageError?
         pytest.param(
             ["--branch=main", "--tag=v1.2.0"],
+            2,
             marks=pytest.mark.xfail(raises=click.UsageError),
         ),
     ],
 )
-def test_dl(message_ix_cli, opts, tmp_path):
-    if isinstance(opts, list):
-        r = message_ix_cli("dl", *opts, str(tmp_path))
-        # TODO Why is this not actually raising a click.UsageError?
-        assert r.exit_code == 2, (r.exception, r.output)
-    else:
-        r = message_ix_cli("dl", opts, str(tmp_path))
+def test_dl(message_ix_cli, opts, exit_code, tmp_path):
+    r = message_ix_cli("dl", *opts, str(tmp_path))
+    assert r.exit_code == exit_code, (r.exception, r.output)
 
-        assert r.exit_code == 0, (r.exception, r.output)
+    if opts == "":
+        # Guess what the latest release will be from GitHub, using __version__.
+        # This string is provided by setuptools-scm based on the most recent
+        # Git tag, e.g. if the tag is 'v3.0.0' it may be '3.0.1.devN+etc'.
+        major = message_ix.__version__.split(".")[0]
 
-        if opts == "":
-            # Guess what the latest release will be from GitHub, using __version__.
-            # This string is provided by setuptools-scm based on the most recent
-            # Git tag, e.g. if the tag is 'v3.0.0' it may be '3.0.1.devN+etc'.
-            major = message_ix.__version__.split(".")[0]
-
-            # 'message-ix dl' defaults to the latest release
-            pattern = re.compile(rf"Default: latest release v{major}\.\d+\.\d+")
-            assert pattern.match(r.output)
+        # 'message-ix dl' defaults to the latest release
+        pattern = re.compile(rf"Default: latest release v{major}\.\d+\.\d+")
+        assert pattern.match(r.output)

--- a/message_ix/tests/test_cli.py
+++ b/message_ix/tests/test_cli.py
@@ -1,5 +1,4 @@
 import re
-import shutil
 from pathlib import Path
 
 import pytest
@@ -8,7 +7,9 @@ import message_ix
 from message_ix import config
 
 
-def test_copy_model(monkeypatch, message_ix_cli, tmp_path, tmp_env, request):
+def test_copy_model(
+    monkeypatch, message_ix_cli, tmp_path, tmp_env, request, tmp_model_dir
+):
     # Use Pytest monkeypatch fixture; this ensures the original value is restored at the
     # end of the test
     monkeypatch.setattr(
@@ -37,18 +38,13 @@ def test_copy_model(monkeypatch, message_ix_cli, tmp_path, tmp_env, request):
     # Check if specific directory will be skipped
 
     # Create a GAMS runtime directory; these have names like "225a", etc.
-    model_path = Path(message_ix.__file__).parent.joinpath(
-        "model", request.node.name, "225c"
-    )
-    model_path.mkdir(parents=True)
+    model_path = tmp_model_dir.joinpath("225c")
+    model_path.mkdir()
 
-    message_ix_cli("copy-model", str(tmp_path))
+    message_ix_cli("copy-model", "--source-dir", tmp_model_dir, f"{tmp_path}-dest")
 
     # Directory is ignored
-    assert not Path(tmp_path / f"{request.node.name}/225c").exists()
-
-    # Clean up
-    shutil.rmtree(model_path.parent)
+    assert not Path(f"{tmp_path}-dest/225c").exists()
 
 
 @pytest.mark.parametrize(

--- a/message_ix/tests/test_cli.py
+++ b/message_ix/tests/test_cli.py
@@ -1,6 +1,7 @@
 import re
 from pathlib import Path
 
+import click
 import pytest
 
 import message_ix
@@ -54,20 +55,30 @@ def test_copy_model(
         "--branch=main",
         "--tag=v1.2.0",
         # Nonexistent tag
-        pytest.param("--tag=v999", marks=pytest.mark.xfail(raises=AssertionError)),
+        pytest.param("--tag=v999", marks=pytest.mark.xfail(raises=ValueError)),
+        # Can't use both --tag and --branch
+        pytest.param(
+            ["--branch=main", "--tag=v1.2.0"],
+            marks=pytest.mark.xfail(raises=click.UsageError),
+        ),
     ],
 )
 def test_dl(message_ix_cli, opts, tmp_path):
-    r = message_ix_cli("dl", opts, str(tmp_path))
+    if isinstance(opts, list):
+        r = message_ix_cli("dl", *opts, str(tmp_path))
+        # TODO Why is this not actually raising a click.UsageError?
+        assert r.exit_code == 2, (r.exception, r.output)
+    else:
+        r = message_ix_cli("dl", opts, str(tmp_path))
 
-    assert r.exit_code == 0, (r.exception, r.output)
+        assert r.exit_code == 0, (r.exception, r.output)
 
-    if opts == "":
-        # Guess what the latest release will be from GitHub, using __version__.
-        # This string is provided by setuptools-scm based on the most recent
-        # Git tag, e.g. if the tag is 'v3.0.0' it may be '3.0.1.devN+etc'.
-        major = message_ix.__version__.split(".")[0]
+        if opts == "":
+            # Guess what the latest release will be from GitHub, using __version__.
+            # This string is provided by setuptools-scm based on the most recent
+            # Git tag, e.g. if the tag is 'v3.0.0' it may be '3.0.1.devN+etc'.
+            major = message_ix.__version__.split(".")[0]
 
-        # 'message-ix dl' defaults to the latest release
-        pattern = re.compile(rf"Default: latest release v{major}\.\d+\.\d+")
-        assert pattern.match(r.output)
+            # 'message-ix dl' defaults to the latest release
+            pattern = re.compile(rf"Default: latest release v{major}\.\d+\.\d+")
+            assert pattern.match(r.output)

--- a/message_ix/tests/test_cli.py
+++ b/message_ix/tests/test_cli.py
@@ -8,7 +8,7 @@ import message_ix
 from message_ix import config
 
 
-def test_copy_model(monkeypatch, message_ix_cli, tmp_path, tmp_env):
+def test_copy_model(monkeypatch, message_ix_cli, tmp_path, tmp_env, request):
     # Use Pytest monkeypatch fixture; this ensures the original value is restored at the
     # end of the test
     monkeypatch.setattr(
@@ -16,7 +16,7 @@ def test_copy_model(monkeypatch, message_ix_cli, tmp_path, tmp_env):
     )
 
     r = message_ix_cli("copy-model", str(tmp_path))
-    assert r.exit_code == 0
+    assert r.exit_code == 0, (r.exception, r.output)
 
     # Copying again without --overwrite fails
     r = message_ix_cli("copy-model", str(tmp_path))
@@ -37,14 +37,15 @@ def test_copy_model(monkeypatch, message_ix_cli, tmp_path, tmp_env):
     # Check if specific directory will be skipped
 
     # Create a GAMS runtime directory; these have names like "225a", etc.
-    model_path = Path(message_ix.__file__).parent.joinpath("model", "225c", "test.txt")
-    model_path.parent.mkdir(parents=True, exist_ok=True)
-    model_path.write_text("foo")
+    model_path = Path(message_ix.__file__).parent.joinpath(
+        "model", request.node.name, "225c"
+    )
+    model_path.mkdir(parents=True)
 
     message_ix_cli("copy-model", str(tmp_path))
 
     # Directory is ignored
-    assert not Path(tmp_path / "225c").exists()
+    assert not Path(tmp_path / f"{request.node.name}/225c").exists()
 
     # Clean up
     shutil.rmtree(model_path.parent)

--- a/message_ix/tests/test_core.py
+++ b/message_ix/tests/test_core.py
@@ -129,8 +129,8 @@ class TestScenario:
             assert "'ixmp'.'3-" in result.stdout.decode()
 
 
-def test_year_int(test_mp):
-    scen = make_dantzig(test_mp, solve=True, multi_year=True)
+def test_year_int(test_mp, request):
+    scen = make_dantzig(test_mp, solve=True, multi_year=True, request=request)
 
     # Dimensions indexed by 'year' are returned as integers for all item types
     assert scen.set("cat_year").dtypes["year"] == "int"

--- a/message_ix/tests/test_feature_bound_activity_shares.py
+++ b/message_ix/tests/test_feature_bound_activity_shares.py
@@ -20,7 +20,7 @@ def calculate_activity(scen, tec="transport_from_seattle") -> pd.Series:
     return scen.var("ACT").groupby(["technology", "mode"])["lvl"].sum().loc[tec]
 
 
-def test_add_bound_activity_up(request, test_mp):
+def test_add_bound_activity_up(request, test_mp, testrun_uid):
     scen = make_dantzig(test_mp, request=request, solve=True, quiet=True)
 
     # data for act bound
@@ -38,7 +38,9 @@ def test_add_bound_activity_up(request, test_mp):
     )
 
     # test limiting one mode
-    clone = scen.clone(scenario=f"{scen.scenario} cloned", keep_solution=False)
+    clone = scen.clone(
+        scenario=f"{scen.scenario}-{testrun_uid} cloned", keep_solution=False
+    )
 
     with clone.transact():
         clone.add_par("bound_activity_up", data)
@@ -96,7 +98,9 @@ def assert_dantzig_solution(s: "Scenario", lp_method: int) -> None:
     "constraint_value",
     [pytest.param(299, marks=pytest.mark.xfail(raises=ModelError)), 301, 325],
 )
-def test_b_a_u_all_modes(request, test_mp, tmp_model_dir, lp_method, constraint_value):
+def test_b_a_u_all_modes(
+    request, test_mp, tmp_model_dir, lp_method, constraint_value, testrun_uid
+):
     """Test ``bound_activity_up`` values applied mode="all".
 
     - In the unconstrained Dantzig problem:
@@ -132,7 +136,9 @@ def test_b_a_u_all_modes(request, test_mp, tmp_model_dir, lp_method, constraint_
     # Ensure the solution is as expected given the LP method
     assert_dantzig_solution(scen, lp_method)
 
-    clone = scen.clone(scenario=f"{scen.scenario} cloned", keep_solution=False)
+    clone = scen.clone(
+        scenario=f"{scen.scenario}-{testrun_uid} cloned", keep_solution=False
+    )
     with clone.transact("Bound all modes of t=transport_from_seattle"):
         name = "bound_activity_up"
         clone.add_par(
@@ -166,7 +172,7 @@ def test_b_a_u_all_modes(request, test_mp, tmp_model_dir, lp_method, constraint_
         assert_equal(calculate_activity(scen).sum(), calculate_activity(clone).sum())
 
 
-def test_commodity_share_up(request, test_mp):
+def test_commodity_share_up(test_mp, request, testrun_uid):
     """Original Solution
     ----------------
 
@@ -250,7 +256,9 @@ def test_commodity_share_up(request, test_mp):
         commodity="cases",
         level="supply",
     )
-    clone = scen.clone(scenario=f"{scen.scenario} share_mode_list", keep_solution=False)
+    clone = scen.clone(
+        scenario=f"{scen.scenario}-{testrun_uid} share_mode_list", keep_solution=False
+    )
     add_data(clone, map_df)
     clone.solve(quiet=True)
 
@@ -266,7 +274,7 @@ def test_commodity_share_up(request, test_mp):
     # add share constraints with mode == 'all'
     map_df2 = map_df.assign(mode="all")
     clone2 = scen.clone(
-        scenario=f"{scen.scenario} share_all_modes", keep_solution=False
+        scenario=f"{scen.scenario}-{testrun_uid} share_all_modes", keep_solution=False
     )
     add_data(clone2, map_df2)
     clone2.solve(quiet=True)
@@ -284,7 +292,7 @@ def test_commodity_share_up(request, test_mp):
     assert new_obj >= orig_obj
 
 
-def test_commodity_share_lo(request, test_mp):
+def test_commodity_share_lo(test_mp, request, testrun_uid):
     scen = make_dantzig(test_mp, request=request, solve=True, quiet=True)
 
     # data for share bound
@@ -297,7 +305,9 @@ def test_commodity_share_lo(request, test_mp):
     exp = 1.0 * calc_share(scen)
 
     # add share constraints
-    clone = scen.clone(keep_solution=False)
+    clone = scen.clone(
+        scenario=f"{scen.scenario}-{testrun_uid} cloned", keep_solution=False
+    )
 
     with clone.transact("Add share constraints"):
         clone.add_set("shares", "test-share")
@@ -339,7 +349,7 @@ def test_commodity_share_lo(request, test_mp):
     assert clone.var("OBJ")["lvl"] >= scen.var("OBJ")["lvl"]
 
 
-def test_add_share_mode_up(request, test_mp):
+def test_add_share_mode_up(request, test_mp, testrun_uid):
     scen = make_dantzig(test_mp, request=request, solve=True, quiet=True)
 
     # data for share bound
@@ -351,7 +361,9 @@ def test_add_share_mode_up(request, test_mp):
     exp = 0.95 * calc_share(scen)
 
     # add share constraints
-    clone = scen.clone(scenario=f"{scen.scenario} share_mode_up", keep_solution=False)
+    clone = scen.clone(
+        scenario=f"{scen.scenario}-{testrun_uid} cloned", keep_solution=False
+    )
 
     with clone.transact("Add share_mode_up"):
         clone.add_set("shares", "test-share")
@@ -379,7 +391,7 @@ def test_add_share_mode_up(request, test_mp):
     assert new_obj >= orig_obj
 
 
-def test_add_share_mode_lo(request, test_mp):
+def test_add_share_mode_lo(request, test_mp, testrun_uid):
     scen = make_dantzig(test_mp, solve=True, request=request, quiet=True)
 
     # data for share bound
@@ -391,7 +403,9 @@ def test_add_share_mode_lo(request, test_mp):
     exp = 1.05 * calc_share(scen)
 
     # add share constraints
-    clone = scen.clone(scenario=f"{scen.scenario} cloned", keep_solution=False)
+    clone = scen.clone(
+        scenario=f"{scen.scenario}-{testrun_uid} cloned", keep_solution=False
+    )
     with clone.transact():
         clone.add_set("shares", "test-share")
         clone.add_par(

--- a/message_ix/tests/test_feature_bound_activity_shares.py
+++ b/message_ix/tests/test_feature_bound_activity_shares.py
@@ -98,9 +98,7 @@ def assert_dantzig_solution(s: "Scenario", lp_method: int) -> None:
     "constraint_value",
     [pytest.param(299, marks=pytest.mark.xfail(raises=ModelError)), 301, 325],
 )
-def test_b_a_u_all_modes(
-    request, test_mp, tmp_model_dir, lp_method, constraint_value, testrun_uid
-):
+def test_b_a_u_all_modes(request, test_mp, tmp_model_dir, lp_method, constraint_value):
     """Test ``bound_activity_up`` values applied mode="all".
 
     - In the unconstrained Dantzig problem:
@@ -136,9 +134,7 @@ def test_b_a_u_all_modes(
     # Ensure the solution is as expected given the LP method
     assert_dantzig_solution(scen, lp_method)
 
-    clone = scen.clone(
-        scenario=f"{scen.scenario}-{testrun_uid} cloned", keep_solution=False
-    )
+    clone = scen.clone(scenario=f"{scen.scenario} cloned", keep_solution=False)
     with clone.transact("Bound all modes of t=transport_from_seattle"):
         name = "bound_activity_up"
         clone.add_par(

--- a/message_ix/tests/test_feature_bound_activity_shares.py
+++ b/message_ix/tests/test_feature_bound_activity_shares.py
@@ -91,10 +91,10 @@ def assert_dantzig_solution(s: "Scenario", lp_method: int) -> None:
     assert_frame_equal(exp, s.var("ACT")[cols])
 
 
-@pytest.mark.parametrize("lp_method", (1, 2, 3, 4, 5))
+@pytest.mark.parametrize("lp_method", [1, 2, 3, 4, 5])
 @pytest.mark.parametrize(
     "constraint_value",
-    (pytest.param(299, marks=pytest.mark.xfail(raises=ModelError)), 301, 325),
+    [pytest.param(299, marks=pytest.mark.xfail(raises=ModelError)), 301, 325],
 )
 def test_b_a_u_all_modes(request, test_mp, tmp_model_dir, lp_method, constraint_value):
     """Test ``bound_activity_up`` values applied mode="all".

--- a/message_ix/tests/test_feature_storage.py
+++ b/message_ix/tests/test_feature_storage.py
@@ -251,9 +251,9 @@ def test_storage(test_mp):
     storage_setup(test_mp, time_duration, "_unequal_time")
 
 
-def test_structure(caplog, test_mp):
+def test_structure(caplog, test_mp, request):
     """:meth:`MESSAGE.initialize` and :meth:`MESSAGE.enforce` handle old structure."""
-    scen = make_dantzig(test_mp)
+    scen = make_dantzig(test_mp, request=request)
 
     # Item name to use for the tests, a parameter, and its dimensions
     name = "storage_initial"

--- a/message_ix/tests/test_macro.py
+++ b/message_ix/tests/test_macro.py
@@ -31,9 +31,11 @@ def w_data(w_data_path):
 
 
 @pytest.fixture(scope="module")
-def _ws(test_mp):
+def _ws(test_mp, request):
     """Reusable fixture with an instance of the Westeros model."""
-    scenario = make_westeros(test_mp, quiet=True).clone(scenario="test_macro")
+    scenario = make_westeros(test_mp, quiet=True, request=request).clone(
+        scenario=f"{request.node.name}_test_macro"
+    )
     scenario.solve()
     yield scenario
 

--- a/message_ix/tests/test_report.py
+++ b/message_ix/tests/test_report.py
@@ -90,8 +90,8 @@ def test_reporter_from_scenario(message_test_mp):
     assert_qty_equal(vom, rep.get(vom_key))
 
 
-def test_reporter_from_dantzig(test_mp):
-    scen = make_dantzig(test_mp, solve=True, quiet=True)
+def test_reporter_from_dantzig(test_mp, request):
+    scen = make_dantzig(test_mp, solve=True, quiet=True, request=request)
 
     # Reporter.from_scenario can handle Dantzig example model
     rep = Reporter.from_scenario(scen)
@@ -100,8 +100,10 @@ def test_reporter_from_dantzig(test_mp):
     rep.get("all")
 
 
-def test_reporter_from_westeros(test_mp):
-    scen = make_westeros(test_mp, emissions=True, solve=True, quiet=True)
+def test_reporter_from_westeros(test_mp, request):
+    scen = make_westeros(
+        test_mp, emissions=True, solve=True, quiet=True, request=request
+    )
 
     # Reporter.from_scenario can handle Westeros example model
     rep = Reporter.from_scenario(scen)

--- a/message_ix/tests/test_util.py
+++ b/message_ix/tests/test_util.py
@@ -46,16 +46,16 @@ def test_make_df_deprecated():
     pdt.assert_frame_equal(pd.DataFrame.from_dict(base), exp)
 
 
-def test_testing_make_scenario(test_mp):
+def test_testing_make_scenario(test_mp, request):
     """Check the model creation functions in message_ix.testing."""
     # MESSAGE-scheme Dantzig problem can be created
-    scen = make_dantzig(test_mp, True)
+    scen = make_dantzig(test_mp, True, request=request)
     assert isinstance(scen, Scenario)
 
     # Multi-year variant can be created
-    scen = make_dantzig(test_mp, solve=True, multi_year=True)
+    scen = make_dantzig(test_mp, solve=True, multi_year=True, request=request)
     assert isinstance(scen, Scenario)
 
     # Westeros model can be created
-    scen = make_westeros(test_mp, solve=True)
+    scen = make_westeros(test_mp, solve=True, request=request)
     assert isinstance(scen, Scenario)

--- a/message_ix/tests/tools/test_add_year.py
+++ b/message_ix/tests/tools/test_add_year.py
@@ -30,7 +30,7 @@ def base_scen_mp(test_mp, request):
             scen.add_par("output", tec_specs + output_specs, 1, "-")
             scen.add_par("var_cost", tec_specs + ["year"], value, "USD/GWa")
 
-    scen.solve(case="original_years", quiet=True)
+    scen.solve(case=f"{request.node.name}_original_years", quiet=True)
 
     yield scen, test_mp
 

--- a/message_ix/util/__init__.py
+++ b/message_ix/util/__init__.py
@@ -27,14 +27,11 @@ def copy_model(
     quiet : bool
         If :any:`False`, print actions to stdout; otherwise display nothing.
     """
-    try:
-        from importlib.resources import as_file, files
-    except ImportError:  # Python < 3.9
-        from importlib_resources import as_file, files
-
     from shutil import copyfile
 
     import ixmp
+
+    import message_ix
 
     path = Path(path).resolve()
 
@@ -46,14 +43,7 @@ def copy_model(
     else:
         output = print
 
-    # Identify the source directory using importlib.resources
-    for traversable in files("message_ix").iterdir():
-        if traversable.name == "model":
-            # Record the pathlib.Path associated with the Traversible object
-            # NB This may not work if `traversable` is, for instance, in a ZIP archive
-            with as_file(traversable) as f:
-                src_dir: Path = f
-            break
+    src_dir: Path = Path(message_ix.__file__).resolve().parent
 
     # Iterate over files in `src_dir`
     for src in src_dir.rglob("*"):

--- a/message_ix/util/__init__.py
+++ b/message_ix/util/__init__.py
@@ -4,6 +4,7 @@ import warnings
 from collections import ChainMap, defaultdict
 from collections.abc import Mapping
 from pathlib import Path
+from typing import Optional
 
 import pandas as pd
 from ixmp.backend import ItemType
@@ -14,12 +15,18 @@ from message_ix.models import MACRO, MESSAGE
 
 
 def copy_model(
-    path: Path, overwrite: bool = False, set_default: bool = False, quiet: bool = False
+    path: Path,
+    source_dir: Optional[Path] = None,
+    overwrite: bool = False,
+    set_default: bool = False,
+    quiet: bool = False,
 ) -> None:
     """Copy the MESSAGE GAMS files to a new `path`.
 
     Parameters
     ----------
+    source_dir : Path
+        If given, copy model files from this directory instead of the main one.
     overwrite : bool
         If :any:`True`, overwrite existing files.
     set_default : bool
@@ -43,7 +50,11 @@ def copy_model(
     else:
         output = print
 
-    src_dir: Path = Path(message_ix.__file__).resolve().parent
+    src_dir = (
+        source_dir
+        if source_dir
+        else Path(message_ix.__file__).resolve().parent.joinpath("model")
+    )
 
     # Iterate over files in `src_dir`
     for src in src_dir.rglob("*"):

--- a/message_ix/util/__init__.py
+++ b/message_ix/util/__init__.py
@@ -16,23 +16,24 @@ from message_ix.models import MACRO, MESSAGE
 
 def copy_model(
     path: Path,
-    source_dir: Optional[Path] = None,
     overwrite: bool = False,
     set_default: bool = False,
     quiet: bool = False,
+    *,
+    source_dir: Optional[Path] = None,
 ) -> None:
     """Copy the MESSAGE GAMS files to a new `path`.
 
     Parameters
     ----------
-    source_dir : Path
-        If given, copy model files from this directory instead of the main one.
     overwrite : bool
         If :any:`True`, overwrite existing files.
     set_default : bool
         If :any:`True`, update the ixmp configuration setting "message model dir".
     quiet : bool
         If :any:`False`, print actions to stdout; otherwise display nothing.
+    source_dir : Path, optional
+        If given, copy model files from this directory instead of the main one.
     """
     from shutil import copyfile
 

--- a/message_ix/util/__init__.py
+++ b/message_ix/util/__init__.py
@@ -62,7 +62,7 @@ def copy_model(
             return False
         return True
 
-    paths = list(filter(_exclude_path, src_dir.rglob("*")))
+    paths = filter(_exclude_path, list(src_dir.rglob("*")))
 
     # Iterate over pre-filtered paths in `src_dir`
     for original_path in paths:

--- a/message_ix/util/__init__.py
+++ b/message_ix/util/__init__.py
@@ -56,19 +56,23 @@ def copy_model(
         else Path(message_ix.__file__).resolve().parent.joinpath("model")
     )
 
-    # Iterate over files in `src_dir`
-    for src in src_dir.rglob("*"):
+    def _exclude_path(p: Path) -> bool:
         # Skip certain files
-        if src.suffix in (".gdx", ".log", ".lst") or re.search("225[a-z]+", str(src)):
-            continue
+        if p.suffix in (".gdx", ".log", ".lst") or re.search("225[a-z]+", str(p)):
+            return False
+        return True
 
+    paths = list(filter(_exclude_path, src_dir.rglob("*")))
+
+    # Iterate over pre-filtered paths in `src_dir`
+    for original_path in paths:
         # Construct the destination path
-        dst = path / src.relative_to(src_dir)
+        dst = path / original_path.relative_to(src_dir)
 
         # Create parent directory
         dst.parent.mkdir(parents=True, exist_ok=True)
 
-        if src.is_dir():
+        if original_path.is_dir():
             # No further action for directories
             continue
 
@@ -84,7 +88,7 @@ def copy_model(
         else:
             output(f"Copy to {dst}")
 
-        copyfile(src, dst)
+        copyfile(original_path, dst)
 
     if set_default:
         ixmp.config.set("message model dir", path)

